### ISSUE TITLE
Support for Session destruction notifications

### ISSUE THREAD

**Issue by gideonred** [replayio/devtools#7003]
The backend will soon support the ability to notify clients of session destructions and give clients the ability to possibly cancel them: https://github.com/RecordReplay/backend/pull/5765

It would be useful for our end users to see these notifications in devtools so that they can distinguish between sessions crashing unexpectedly on them, versus session we are trying to terminate in a controlled manner.

**Comment by jaril** [comment]
@jonbell-lot23 this might be a fun one to get your eyes on. There's a bunch of information to follow by backtracking to that backend issue and leading the related code/comments/issues but here are my notes:

Sessions get destroyed. There are times when these get destroyed because it's been idle for some time, and so there's no user impact. But sometimes, they get destroyed while a user is actively using it. That's when the user gets some error that requires them to refresh their browser.

The problem is that in the cases where there are user impact, there's a difference between 1) an actual error vs 2) us destroying a session in a controlled manner (e.g., in response to a deploy or EC2 spot termination). In the past, both of those would lead to error screens.

Now we have some capability to tell 1 from 2. And if it's 2, the frontend now is informed that their client's going to be destroyed soon:tm:.

So some new implications:
a) We can message 1 differently from 2. 1 is an actual unexpected session error. 2 is, theoretically, an expected session error, if you can even call it an error.
b) For case 2, we can take some input from the user. The user can say whoop you know what I'd prefer if you shut it down later, and the backend can push that pending shutdown off.

I know the capability to do [a was already shipped](https://github.com/replayio/backend/pull/5765). But it looks like b is a follow up.

**tldr; it's like you have dial up internet and somebody's about to use the phone. Before, they'd just pick it up and ruin your internet. Now they'll yell at you to give you a heads up that your internet's about to be cut off in t minutes. And potentially, you could yell back "GIVE ME FIVE MINUTES I JUST NEED TO GO TO A SAVE POINT" so you don't get too interrupted.**

@gideonred Did I get that right?

**Comment by jonbell-lot23** [comment]
Super cool, more information in these situations can really help.

Off the top of my head, I'm not sure if we want to be as literal as allowing a person to pause a session from being destroyed. It's a classic example of "don't make me think" where we should probably just use this data to do the right thing. And I think we know what "the right thing" is.

If we haven't detected activity from the user for a few minutes, maybe we just destroy the session without having to ask. (Which is what happens today) And if we do detect activity, maybe we hold off. (Which is what anyone would want, without a prompt)

**Comment by loganfsmyth** [comment]
> A lot (~40%) of our sessions are destroyed because we choose to, because of a deployment.

That statistic is pretty misleading in my opinion because deployments are the _only_ time we force-destroy a session for a reason that isn't an actual recording error, so any session that exists when we deploy will be counted toward that even if the session wasn't being actively used.

@jonbell-lot23 
To clarify the details a bit here now, the API will inform the frontend of a few basic things:
* Give a warning if the API has not seen any activity recently and will destroy the session if there continues to be no activity (referred to as `mayDestroy`).
* Give a warning that there is a deployment in progress and the session will be destroyed (refered to as `willDestroy`). We currently send this about 25 minutes before the session will be destroyed, but that's more a best-effort.

Keep in mind that these can both happen independently.

My thought for behavior here was something along the lines of:
1. If the frontend gets a `willDestroy` notification it means we're deploying and _definitely_ going to destroy the session, so I was thinking in that case, it may still be a way away time-wise, so:
  * If it's more than like 5 minutes in the future, the frontend could quietly keep track of the fact that it knows the session will be force-destroyed, and take that into account when showing the user warnings for `mayDestroy` (or wait til the 5-minute mark to do the next point)
  * If we're close to the force-destroy time, and we haven't already shown the user any other warning from `mayDestroy`, show an explicit warning that the user needs to refresh the page very soon.
2. If the frontend gets a `mayDestroy` notification because of API inactivity
  * if it has seen user interaction in like the last 5-10 minutes or something, automatically cancel the pending `mayDestroy`. If there is a pending `willDestroy` too, maybe do the next point instead
  * If no recent user interaction, show a popup telling the user we intend to destroy the session, but they may cancel it. I'm imagining like Netflix's "are you still there" type things. If there is an upcoming willDestroy _too_, probably let the user know in this popup too to encourage them to refresh instead of cancelling the `mayDestroy`.

**Comment by loganfsmyth** [comment]
As part of this work, we also need to fix our session-destroy dialog to give the user the message our API returned, because right now it appears we fall back to a generic error. Even if the backend gave a message like "Session died due to inactivity", the UI shows essentially "Something went wrong, we'll look into it" (see https://github.com/replayio/backend/issues/5856)

**Comment by jonbell-lot23** [comment]
Thanks for the background! It's Saturday and I'm just skimming but wanted to respond so I don't block anyone.

This background is great and I understand more of the nuance now. I'm pretty sure we're basically just down to messaging differently based on what happened. So I see a few insights:

1. Don't make me think -- just do stuff as long as the visitor doesn't need to know about it. For example, if something is going to happen in 30 minutes, we absolutely don't need to tell them yet.
2. Message clearly when we need to message -- when everything blows up, obviously a user needs to know why, and have a clear way forward. So whether the message is "Our apologies, we just deployed, but reload and we'll send you on your way" or "Whoa, fatal error and we're working on it" I think both messages are good and helpful.

It is possible, due to being Saturday and me skimming, that I am missing details. But I think we're aligned here. Yes to contextual messaging when it's important, and no to over-messaging when something is going to happen in the future.

**Comment by loganfsmyth** [comment]
The main thing in my mind is that there is absolutely value in warning the user before we destroy a session, and currently the UX around that is undefined.

* If we're deploying, there is value in messaging when we are close to the time when expect to destroy a session without recourse, because they can still take action to decide that it is a good time to reload, so they can avoid getting any error at all, and more importantly so they have time to I note what they were doing. Maybe they had some JS evaluated in their console. That would be gone if they refresh, so potentially we want to warn them ahead of time to save that. In a perfect world we could automatically do that, but right now we don't and that's a much larger and more difficult project.
* Separately, even for destroying the session due to inactivity, there is value in giving them some amount of time to keep the session alive, because they might still be debugging, just doing so in their head, or by skimming code in their editor instead of Replay, but if they were driving that debugging process with some series of console logs in the replay window and they suddenly lose all of that, it would be quite frustrating.

**Comment by jonbell-lot23** [comment]
That's reasonable.

So I'd imitate what command-line computing has done for many decades, the whole "Hey the server is shutting down in x minutes" thing.

I still have concerns about forecasting it too far in the future, so here's what I'd propose:

* We tell the person when it's gonna happen in the next five minutes
* I think it's reasonable to explore a banner across the top the full app. Copy to come, but just letting them know that sometime in the next five minutes would be a great time to reload is helpful

The counter-arguments and my responses to both of these points are:

* We should give them more time: the banner has a drawback of its own because it's quite loud. I think five minutes finds a good balance between having enough time without being too in-your-face.
* The banner is too bold: this is the flip side of the previous item. If it's important enough to say, it's important enough to say clearly. I wouldn't say the banner should live there for 30 minutes, but when it's down to five minutes, I think it's appropriate.

Thoughts?

**Comment by loganfsmyth** [comment]
Sounds great, and yes I agree that the banner shouldn't be there for a super long time, 5 minutes or even less is fine with me. 5 minute sounds good for a forced-deploy since they have to refresh and giving them time to plan for that is important, and a banner or something feels like a good way to do that. For the case of inactivity, a banner is an option, though I do wonder if it'd be even better to just show like a dim overlay with a warning that clears itself as soon as they move their mouse on the window or something, in which case 5 minutes might be even longer than needed.

**Comment by jonbell-lot23** [comment]
Going to assign myself for a quick banner design, but don't wait on me. Obviously the main work is hooking up the code, and a placeholder banner is fine while I work on other priorities.

**Comment by jaril** [comment]
+1 on direction. Unsure about something as loud as a banner, but at the same time, it's so in your face that we'll run into it and have a better sense of how to tone it down once it's in.

@loganfsmyth my understanding is that in the `willDestroy` case, refreshing the page should start a new session and the user can keep debugging. Is that right??

**Comment by loganfsmyth** [comment]
> @loganfsmyth my understanding is that in the willDestroy case, refreshing the page should start a new session and the user can keep debugging. Is that right??

Yep that's right!

**Comment by jonbell-lot23** [comment]
> +1 on direction. Unsure about something as loud as a banner, but at the same time, it's so in your face that we'll run into it and have a better sense of how to tone it down once it's in.

Also notable that we're looking at two versions of loud:

1. The space it takes up
2. The actual colors used

A banner is a great way to be obvious but polite. You can use a white background (rather than hot pink), you can make it not super tall (rather than taking half the screen), and it doesn't have to put you in a new mode that requires dismissing (as opposed to a modal dialog).

We have a similar banner in the browser. A simple one-liner with neutral colors is easy to see but not disruptive.

**Comment by gideonred** [comment]
This ticket has been moved to [FE-203](https://linear.app/replay/issue/FE-203/support-for-session-destruction-notifications)
